### PR TITLE
Create smartlock-key controllers and integration tests

### PR DESCRIPTION
**Pull request by salvanicola**

### CODE REVIEW

File: web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs [hunk 1]
```diff
@@ -1,32 +1,54 @@
...
+		_smartLockKeyRepository = smartLockKeyRepository;
+		_nukiSmartLockRepository = nukiSmartLockRepository;
+		_nukiCredentialRepository = nukiCredentialRepository;
 	}
 
-	public async Task<Result<CreateSmartLockKeyPresentationResponse>> Handle(CreateSmartLockKeyPresentationRequest presentationRequest)
+	public async Task<Result<SmartLockKeyDto>> Handle(CreateSmartLockKeyParams request)
 	{
-		var generatedKeyResult = _smartLockKeyGeneratorManager.GenerateSmartLockKey(presentationRequest.SmartLockId, presentationRequest.ExpiryDate);
+		// TODO switch on provider or move logic on another interactor
```
Comment (vash72) [inline web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs:30]:
```suggestion

```
can we check if the provider is not "nuki" and throw "UnsupportedSmartLockProvider" ?

**Review by vash72 — CHANGES_REQUESTED** [review]

File: web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	
+	public required DateTime CreationDate { get; set; }
+	
+	public required DateTime ExpiryDate { get; set; }
+
+	public required string Token { get; set; }
+
+	public bool IsDisabled { get; set; }
+	
+	public int UsageCounter { get; set; }
```
Comment (vash72) [inline web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs:15]:
```suggestion
	public required  int UsageCounter { get; set; }
```

File: web-api/tests/Kerbero.Integration.Tests/SmartLockKeys/CreateSmartLockKeyIntegrationTests.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+public class CreateSmartLockKeyIntegrationTests: IDisposable
+{
+	private readonly KerberoWebApplicationFactory<Program> _application;
+	private readonly HttpTest _httpTest;
+
+	public CreateSmartLockKeyIntegrationTests()
+	{
+		_application = new KerberoWebApplicationFactory<Program>();
+		_application.Server.PreserveExecutionContext = true; // fixture for Flurl
```
Comment (vash72) [inline web-api/tests/Kerbero.Integration.Tests/SmartLockKeys/CreateSmartLockKeyIntegrationTests.cs:19]:
```suggestion
		_application.Server.PreserveExecutionContext = true; // fixture for Flurl
```
pleas explain what this line do, so we can understand

**Review by vash72 — CHANGES_REQUESTED** [review]

File: web-api/src/Kerbero.Infrastructure/SmartLockKeys/Mappers/SmartLockKeyMapper.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+public static class SmartLockKeyMapper
+{
+	public static SmartLockKeyEntity Map(SmartLockKeyModel smartLockKeyModel)
+	{
+		return new SmartLockKeyEntity
+		{
+			Token = smartLockKeyModel.Token,
+			CreationDate = smartLockKeyModel.CreationDate.ToUniversalTime(),
+			ExpiryDate = smartLockKeyModel.ExpiryDate.ToUniversalTime(),
```
[the remark this answers is not in the document]
Comment (salvanicola) [reply web-api/src/Kerbero.Infrastructure/SmartLockKeys/Mappers/SmartLockKeyMapper.cs:17]: yes

File: web-api/tests/Kerbero.Integration.Tests/SmartLockKeys/CreateSmartLockKeyIntegrationTests.cs (hunk 3 above)
Comment (salvanicola) [reply web-api/tests/Kerbero.Integration.Tests/SmartLockKeys/CreateSmartLockKeyIntegrationTests.cs:19]: https://github.com/tmenier/Flurl/issues/493#issuecomment-573095892

File: web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs [hunk 5]
```diff
@@ -1,32 +1,58 @@
...
-		var createResult = await _smartLockKeyPersistentRepository.Create(generatedKeyResult.Value);
-		if (createResult.IsFailed)
+		var nukiSmartLock = await _nukiSmartLockRepository.Get(nukiCredentialResult.Value, @params.SmartLockId);
+		
+		if (nukiSmartLock.IsFailed)
+		{
+			return Result.Fail(nukiCredentialResult.Errors);
+		}
+		
+		var generatedKeyResult = SmartLockKeyUtils.CreateKey(@params.SmartLockId, @params.ExpiryDate, @params.CredentialId);
```
Comment (kuamanet) [inline web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs:49]:
```suggestion
		var generatedKeyResult = SmartLockKeyModel.CreateKey(@params.SmartLockId, @params.ExpiryDate, @params.CredentialId);
```
This is a perfect example of business logic that should stay inside the domain entity of this feature

File: web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	
+	public required DateTime CreationDate { get; set; }
+	
+	public required DateTime ExpiryDate { get; set; }
+
+	public required string Token { get; set; }
+
+	public required bool IsDisabled { get; set; }
+	
+	public required int UsageCounter { get; set; }
```
Comment (kuamanet) [inline web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs:15]:
```suggestion
	public bool IsDisabled { get; set; }
	
	public int UsageCounter { get; set; }
```

File: web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+	
+	public required string SmartLockId { get; set; }
+
+	public required int CredentialId { get; set; }
```
Comment (kuamanet) [inline web-api/src/Kerbero.Infrastructure/SmartLockKeys/Entities/SmartLockKeyEntity.cs:19]:
```suggestion
	public int CredentialId { get; set; }
```

File: web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs [hunk 8]
```diff
@@ -1,32 +1,58 @@
...
+		_nukiCredentialRepository = nukiCredentialRepository;
 	}
 
-	public async Task<Result<CreateSmartLockKeyPresentationResponse>> Handle(CreateSmartLockKeyPresentationRequest presentationRequest)
+	public async Task<Result<SmartLockKeyDto>> Handle(CreateSmartLockKeyParams @params)
 	{
-		var generatedKeyResult = _smartLockKeyGeneratorManager.GenerateSmartLockKey(presentationRequest.SmartLockId, presentationRequest.ExpiryDate);
+		if (@params.SmartLockProvider != SmartLockProvider.Nuki)
+		{
+			// TODO return Result.Fail(new UnsupportedSmartLockProviderError()˙);
```
[the remark this answers is not in the document]
Comment (salvanicola) [reply web-api/src/Kerbero.Domain/SmartLockKeys/Interactors/CreateSmartLockKeyInteractor.cs]: The correction too

**Review by kuamanet — APPROVED** [review]

**Pull request merged**